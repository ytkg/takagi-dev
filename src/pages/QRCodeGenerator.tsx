import React, { useState, useRef } from 'react';
import SEO from '../components/SEO';
import QRCode from 'react-qr-code';

export default function QRCodeGenerator() {
  const [text, setText] = useState('');
  const qrCodeRef = useRef<HTMLDivElement>(null);

  const handleSave = () => {
    if (!qrCodeRef.current) {
      return;
    }

    const svgElement = qrCodeRef.current.querySelector('svg');
    if (!svgElement) {
      return;
    }

    const svgString = new XMLSerializer().serializeToString(svgElement);
    const svgDataUri = `data:image/svg+xml;base64,${btoa(svgString)}`;

    const img = new Image();
    img.onload = () => {
      const canvas = document.createElement('canvas');
      canvas.width = img.width;
      canvas.height = img.height;
      const ctx = canvas.getContext('2d');
      if (ctx) {
        ctx.drawImage(img, 0, 0);
        const pngDataUri = canvas.toDataURL('image/png');
        const downloadLink = document.createElement('a');
        downloadLink.href = pngDataUri;
        downloadLink.download = 'qrcode.png';
        document.body.appendChild(downloadLink);
        downloadLink.click();
        document.body.removeChild(downloadLink);
      }
    };
    img.src = svgDataUri;
  };

  return (
    <div className="p-8">
      <SEO
        title="QR Code Generator | takagi.dev"
        description="Generate QR codes instantly from text or URLs."
        path="/tools/qr-code-generator"
        image="/ogp.png"
      />
      <h1 className="text-2xl font-bold mb-4">QR Code Generator</h1>
      <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
        <div>
          <h2 className="text-lg font-semibold mb-2">Input</h2>
          <textarea
            className="w-full h-48 p-2 border rounded-md font-mono text-sm bg-gray-50 dark:bg-gray-800 dark:text-white"
            value={text}
            onChange={(e) => setText(e.target.value)}
            placeholder="Enter text or URL to generate QR code..."
          />
        </div>
        <div>
          <h2 className="text-lg font-semibold mb-2">Output</h2>
          <div
            className="w-full h-48 p-4 border rounded-md bg-white flex items-center justify-center"
            data-testid="qr-code-output"
            ref={qrCodeRef}
          >
            {text ? (
              <QRCode value={text} size={160} viewBox="0 0 256 256" title="qr code" />
            ) : (
              <span className="text-gray-400">QR code will appear here</span>
            )}
          </div>
        </div>
      </div>
      <div className="mt-4 flex space-x-2">
        <button
          onClick={() => setText('')}
          className="px-4 py-2 bg-gray-500 text-white rounded-md hover:bg-gray-600"
          disabled={!text}
        >
          Clear
        </button>
        <button
          onClick={handleSave}
          className="px-4 py-2 bg-blue-500 text-white rounded-md hover:bg-blue-600 disabled:bg-blue-300"
          disabled={!text}
        >
          Save Image
        </button>
      </div>
    </div>
  );
}
