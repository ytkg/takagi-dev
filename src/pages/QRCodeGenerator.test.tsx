import { render, screen, fireEvent } from '@testing-library/react';
import QRCodeGenerator from './QRCodeGenerator';
import '@testing-library/jest-dom';

describe('QRCodeGenerator', () => {
  test('renders the component', () => {
    render(<QRCodeGenerator />);
    expect(screen.getByText('QR Code Generator')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Enter text or URL to generate QR code...')).toBeInTheDocument();
  });

  test('does not display a QR code initially', () => {
    render(<QRCodeGenerator />);
    const outputDiv = screen.getByTestId('qr-code-output');
    expect(outputDiv.querySelector('svg')).not.toBeInTheDocument();
    expect(screen.getByText('QR code will appear here')).toBeInTheDocument();
  });

  test('displays a QR code when text is entered', async () => {
    render(<QRCodeGenerator />);
    const inputArea = screen.getByPlaceholderText('Enter text or URL to generate QR code...');
    fireEvent.change(inputArea, { target: { value: 'Hello World' } });

    const outputDiv = await screen.findByTestId('qr-code-output');
    const qrCodeSvg = outputDiv.querySelector('svg');
    expect(qrCodeSvg).toBeInTheDocument();
    expect(screen.queryByText('QR code will appear here')).not.toBeInTheDocument();
  });

  test('clears the input and QR code when clear button is clicked', async () => {
    render(<QRCodeGenerator />);
    const inputArea = screen.getByPlaceholderText('Enter text or URL to generate QR code...');
    const clearButton = screen.getByText('Clear');
    const outputDiv = screen.getByTestId('qr-code-output');

    // Enter text and verify QR code appears
    fireEvent.change(inputArea, { target: { value: 'Test' } });
    const qrCodeSvg = await screen.findByTestId('qr-code-output').then(div => div.querySelector('svg'));
    expect(qrCodeSvg).toBeInTheDocument();
    expect(inputArea).toHaveValue('Test');

    // Click clear and verify everything is reset
    fireEvent.click(clearButton);
    expect(inputArea).toHaveValue('');
    expect(outputDiv.querySelector('svg')).not.toBeInTheDocument();
    expect(screen.getByText('QR code will appear here')).toBeInTheDocument();
  });

  test('matches the snapshot', () => {
    const { container } = render(<QRCodeGenerator />);
    expect(container).toMatchSnapshot();
  });
});
