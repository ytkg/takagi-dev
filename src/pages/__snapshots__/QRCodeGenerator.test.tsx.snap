// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QRCodeGenerator > matches the snapshot 1`] = `
<div>
  <div
    class="p-8"
  >
    <h1
      class="text-2xl font-bold mb-4"
    >
      QR Code Generator
    </h1>
    <div
      class="grid grid-cols-1 md:grid-cols-2 gap-4"
    >
      <div>
        <h2
          class="text-lg font-semibold mb-2"
        >
          Input
        </h2>
        <textarea
          class="w-full h-48 p-2 border rounded-md font-mono text-sm bg-gray-50 dark:bg-gray-800 dark:text-white"
          placeholder="Enter text or URL to generate QR code..."
        />
      </div>
      <div>
        <h2
          class="text-lg font-semibold mb-2"
        >
          Output
        </h2>
        <div
          class="w-full h-48 p-4 border rounded-md bg-white flex items-center justify-center"
          data-testid="qr-code-output"
        >
          <span
            class="text-gray-400"
          >
            QR code will appear here
          </span>
        </div>
      </div>
    </div>
    <div
      class="mt-4 flex space-x-2"
    >
      <button
        class="px-4 py-2 bg-gray-500 text-white rounded-md hover:bg-gray-600 transition-colors"
        disabled=""
      >
        Clear
      </button>
      <button
        class="px-4 py-2 bg-blue-500 text-white rounded-md hover:bg-blue-600 disabled:bg-blue-300 transition-colors"
        disabled=""
      >
        Save Image
      </button>
    </div>
  </div>
</div>
`;
